File: Java/src/main/java/com/adaptionsoft/games/uglytrivia/RollFactory.java
package com.adaptionsoft.games.uglytrivia;

import java.util.Random;

public class RollFactory extends Base {

    public RollFactory(CurrentPlayer currentPlayer,
            PlayerNames names,
            Places places,
            Purses purses,
            PenaltyBox penaltyBox,
            Categories categories,
            QuestionsByCategory questions,
            boolean[] isGettingOutOfPenaltyBox) {
        super(currentPlayer,
                names,
                places,
                purses,
                penaltyBox,
                categories,
                questions,
                isGettingOutOfPenaltyBox);
    }

    public Action create(Random rand) {
        int roll = calcRoll(rand);
        if (penaltyBox.isIn()) {
            PenaltyBoxRollFactory factory = new PenaltyBoxRollFactory(currentPlayer,
                    names,
                    places,
                    purses,
                    penaltyBox,
                    categories,
                    questions,
                    isGettingOutOfPenaltyBox,
                    roll);
            return factory.create();
        } else {
            return new Roll(currentPlayer,
                    names,
                    places,
                    purses,
                    penaltyBox,
                    categories,
                    questions,
                    isGettingOutOfPenaltyBox,
                    roll);
        }
    }

    private int calcRoll(Random rand) {
        return rand.nextInt(5) + 1;
    }
}

File: Csharp/Trivia/RollFactory.cs
using System;

namespace UglyTrivia
{
    public class RollFactory : Base
    {
        public RollFactory(CurrentPlayer currentPlayer, PlayerNames names, Places places, Purses purses, PenaltyBox penaltyBox, Categories categories, QuestionsByCategory questions, bool[] isGettingOutOfPenaltyBox)
            : base(currentPlayer, names, places, purses, penaltyBox, categories, questions, isGettingOutOfPenaltyBox)
        {
        }

        public Action Create(Random rand)
        {
            int roll = CalcRoll(rand);
            if (penaltyBox.IsIn())
            {
                PenaltyBoxRollFactory factory = new PenaltyBoxRollFactory(currentPlayer, names, places, purses, penaltyBox, categories, questions, isGettingOutOfPenaltyBox, roll);
                return factory.Create();
            }
            else
            {
                return new Roll(currentPlayer, names, places, purses, penaltyBox, categories, questions, isGettingOutOfPenaltyBox, roll);
            }
        }

        private int CalcRoll(Random rand)
        {
            return rand.Next(5) + 1;
        }
    }
}
